python_modules/dagster/dagster/_core/execution/plan/execute_step.py

```diff
@@ -859,18 +896,40 @@ def _log_asset_materialization_events_for_asset(
             f"Unexpected asset execution type {execution_type}",
         )
 
-        yield from (
-            (
-                DagsterEvent.asset_materialization(step_context, materialization)
-                for materialization in _get_output_asset_materializations(
-                    asset_key,
-                    partitions,
-                    output,
-                    output_def,
-                    manager_metadata,
-                    step_context,
+        # This is a temporary workaround to prevent duplicate observation events from external
+        # observable assets that were auto-converted from source assets. These assets yield
+        # observation events through the context in their body, and will continue to do so until we
+        # can convert them to using ObserveResult, which requires a solution to partition-scoped
+        # metadata and data version on output. We identify these auto-converted assets by looking
+        # for OBSERVATION-type asset that doesn't have a data version or metadata, which should only
+        # occur for auto-converted these auto-converted source assets. This can be removed when
+        # source asset observation functions are converted to use ObserveResult.
+        if (
```

Comment:
Can we structure this condition to be more bulletproof?

Right now AFAICT the condition `and SYSTEM_METADATA_KEY_SOURCE_ASSET_OBSERVATION in output.metadata` is meaningful because if is NOT true the behavior is exactly the same since the following `elif` would eval to False and then nothing would happen

Reply:
Good catch that was a bug from a rebase-- the following conditional is now:

```
elif execution_type != AssetExecutionType.UNEXECUTABLE:
```

Which will eval true for observables, which is how it should be. So the following conditional will proceed to auto-generate `AssetObservation`, which we want to skip for assets with this special metadata key.

Reply:
You should add a test that would have failed with the old code

Reply:
Some of the new test_observe_result tests did fail, which is how I caught it